module/zfs/vdev_draid.c

```diff
@@ -0,0 +1,1657 @@
+ * Copyright (c) 2020 by Lawrence Livermore National Security, LLC.
+ */
+
+#include <sys/zfs_context.h>
+#include <sys/spa.h>
+#include <sys/spa_impl.h>
+#include <sys/vdev_impl.h>
+#include <sys/vdev_draid_impl.h>
+#include <sys/vdev_raidz.h>
+#include <sys/vdev_rebuild.h>
+#include <sys/abd.h>
+#include <sys/zio.h>
+#include <sys/nvpair.h>
+#include <sys/zio_checksum.h>
+#include <sys/fs/zfs.h>
+#include <sys/fm/fs/zfs.h>
+
+#ifdef ZFS_DEBUG
+#include <sys/vdev.h>	/* For vdev_xlate() in vdev_draid_io_verify() */
+#endif
+
+/*
+ * dRAID is a distributed parity implementation for ZFS. A dRAID vdev is
+ * comprised of multiple raidz redundancy groups which are spread over the
+ * dRAID children. To ensure an even distribution, and avoid hot spots, a
+ * permutation mapping is applied to the order of the dRAID children.
+ * This mixing effectively distributes the parity columns evenly over all
+ * of the disks in the dRAID.
+ *
+ * This is beneficial because it means when resilvering all of the disks
+ * can participate thereby increasing the available IOPs and bandwidth.
+ * Furthermore, by reserving a small fraction of each child's total capacity
+ * virtual distributed spare disks can be created. These spares similarly
+ * benefit from the performance gains of spanning all of the children. The
+ * consequence of which is that resilvering to a distributed spare can
+ * substantially reduce the time required to restore full parity to pool
+ * with a failed disks.
+ *
+ * === dRAID group layout ===
+ *
+ * First, let's define a "row" in the configuration to be a 16M chunk from
+ * each physical drive at the same offset. This is the minimum allowable
+ * size since it must be possible to store a full 16M block when there is
+ * only a single data column. Next, we defined a "group" to be a set of
+ * sequential disks containing both the parity and data columns. We allow
+ * groups to span multiple rows in order to align any group size to any
+ * number of physical drives. Finally, a "slice" is comprised of the rows
+ * which contains the target number of groups. The permutation mappings
+ * are applied in a round robin fashion to each slice.
+ *
+ * Given n drives in a group (including parity drives) and m physical drives
+ * (not including the spare drives), we can distribute the groups across r
+ * rows without remainder by selecting the least common multiple of n and m
+ * as the number groups; i.e. ngroups = LCM(n, m).
+ *
+ * In the example below, there are 17 physical drives in the configuration,
+ * with two drives worth of spare capacity. The group size is 8, there are
+ * 15 groups and 8 rows per group. If the row chunk size is 16M, the slice
+ * chunk size will be 128M.
+ *
+ *
+ *                              data disks                         spares
+ *    +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+ *    | 5 | 11| 7 | 13| 6 | 9 | 12| 1 | 4 | 2 | 0 | 10| 15| 8 | 3 | 14| 16|
```

Comment:
I think this row is the child ID (which physical disk).  Maybe put a horizontal line under this?

Reply:
I've updated to diagram and text to prove what I hope is a better example.